AX: YouTube embed elements should create AX elements based on their node

hoffmanjoshua opened the pull request:
#### aa91363e81fca4caddf839b549de043fc456bab0
<pre>
AX: YouTube embed elements should create AX elements based on their node
<a href="https://bugs.webkit.org/show_bug.cgi?id=296488">https://bugs.webkit.org/show_bug.cgi?id=296488</a>
<a href="https://rdar.apple.com/156331833">rdar://156331833</a>

Reviewed by Tyler Wilcock.

Accessibility elements are (in most cases) based off of the renderer. For
YouTube embed elements, due to some special-casing, the Embed parent and
the shadow dom child share the same renderer (see
HTMLPlugInElement::createElementRenderer). This breaks accessibility
assumptions. Instead, add a rare data field on render objects to indicate
if it is a youtube embed replacement, and if so, default to using its
renderer.

* LayoutTests/accessibility/youtube-embed-accessibility-hierarchy-expected.txt: Added.
* LayoutTests/accessibility/youtube-embed-accessibility-hierarchy.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::get const):
(WebCore::AXObjectCache::getOrCreate):
* Source/WebCore/html/HTMLPlugInElement.cpp:
(WebCore::HTMLPlugInElement::createElementRenderer):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::markIsYouTubeReplacement):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isYouTubeReplacement const):

Canonical link: <a href="https://commits.webkit.org/297939@main">https://commits.webkit.org/297939@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/932d6b15e0c6d6ae183b4831a75f62ecf7652239

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113251 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32991 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23464 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119459 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/64216 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115213 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33639 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41549 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86196 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41442 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116198 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26869 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101875 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66520 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26137 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20003 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63213 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96247 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20078 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122676 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40369 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30087 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95042 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40714 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98090 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94786 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24227 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39935 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17750 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/36465 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40215 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45714 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39856 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43189 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41593 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1010,6 +1014,9 @@ AccessibilityObject* AXObjectCache::getOrCreate(Node& node, IsPartOfRelation isP
 AccessibilityObject* AXObjectCache::getOrCreate(RenderObject& renderer)
 {
+    if (renderer.isYouTubeReplacement())
+        return getOrCreate(renderer.node());
```
Can we add `[[unlikely]]` to this branch to mitigate the cost?

twilco commented on Source/WebCore/rendering/RenderObject.h:
```diff
@@ -684,6 +684,9 @@ class RenderObject : public CachedImageClient {
     bool isExcludedAndPlacedInBorder() const { return isExcludedFromNormalLayout() && isLegend(); }
 
+    bool isYouTubeReplacement() const { return hasRareData() && rareData().isYouTubeReplacement; }
+    void setIsYouTubeReplacement(bool = true);
```
We never pass a `false` value to this. Maybe we can remove the bool parameter and rename it to `markIsYouTubeReplacement()`?

hoffmanjoshua replied:
Fixed!

twilco commented on Source/WebCore/rendering/RenderObject.h:
```diff
@@ -1294,6 +1297,8 @@ class RenderObject : public CachedImageClient {
         // From RenderBox
         RefPtr<ControlPart> controlPart;
+
+        bool isYouTubeReplacement { false };
```
Putting a `bool` after an 8-byte value like a `RefPtr` is probably going to introduce a whole ton of unnecessary padding. Typically it's best to stack small (e.g. 1-byte bools, OptionSets) values together. Running `xcrun python3 Tools/Scripts/dump-class-layout WebCore RenderObjectRareData` can help you find the best place to put this.

hoffmanjoshua replied:
Good call (and awesome tool!)

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -584,9 +584,11 @@ void AXObjectCache::setIsolatedTreeFocusedObject(AccessibilityObject* focus)
-    if (renderID)
-        return m_objects.get(*renderID);
+    if (renderer && !renderer->isYouTubeReplacement()) {
+        auto renderID = renderer ? m_renderObjectMapping.getOptional(*renderer) : std::nullopt;
```
I think this `renderer ?` check is unnecessary since we check that it's non-null to enter this block.

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -981,7 +984,8 @@ AccessibilityObject* AXObjectCache::getOrCreate(Node& node, IsPartOfRelation isP
         bool isPopover = element && element->hasAttributeWithoutSynchronization(popoverAttr);
         bool isAreaElement = is<HTMLAreaElement>(element);
-        if (!inCanvasSubtree && !insideMeterElement && !hasDisplayContents && !isPopover && !isNodeFocused(node) && !isAreaElement)
+        bool replacementWillCreateRenderer = renderer && renderer->isYouTubeReplacement();
```
I wonder if it's worth adding a comment explaining why we have this exception for YouTube replacement things?

hoffmanjoshua replied:
Done!

hoffmanjoshua committed:
AX: YouTube embed elements should create AX elements based on their node
https://bugs.webkit.org/show_bug.cgi?id=296488
rdar://156331833

Reviewed by Tyler Wilcock.

Accessibility elements are (in most cases) based off of the renderer. For
YouTube embed elements, due to some special-casing, the Embed parent and
the shadow dom child share the same renderer (see
HTMLPlugInElement::createElementRenderer). This breaks accessibility
assumptions. Instead, add a rare data field on render objects to indicate
if it is a youtube embed replacement, and if so, default to using its
renderer.

* LayoutTests/accessibility/youtube-embed-accessibility-hierarchy-expected.txt: Added.
* LayoutTests/accessibility/youtube-embed-accessibility-hierarchy.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::get const):
(WebCore::AXObjectCache::getOrCreate):
* Source/WebCore/html/HTMLPlugInElement.cpp:
(WebCore::HTMLPlugInElement::createElementRenderer):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::markIsYouTubeReplacement):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isYouTubeReplacement const):

Canonical link: https://commits.webkit.org/297939@main
